pkg/middleware/batch.go

```diff
@@ -0,0 +1,150 @@
+ * Copyright 2020, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/go-chi/render"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// BatchResposeItem holds the structure for each item
+type BatchResposeItem struct {
+	Status       int         `json:"status"`
+	RequestID    string      `json:"requestID"`
+	Method       string      `json:"method"`
+	URL          string      `json:"url"`
+	OrigResponse interface{} `json:"origResponse"`
+}
+
+// BatchResponse has the structure for the final response
+type BatchResponse struct {
+	StartedAt  time.Time          `json:"startedAt"`
+	EndedAt    time.Time          `json:"endedAt"`
+	ErrorCount int                `json:"errorCount"`
+	Response   []BatchResposeItem `json:"response"`
+}
+
+// BatchWriter implements http.ResponseWriter
+type BatchWriter struct {
+	statusCode int
+	headerMap  http.Header
+	method     string
+	url        string
+
+	BResponse BatchResponse
+}
+
+// WriteHeader sets the status code
+func (rec *BatchWriter) WriteHeader(code int) {
+	rec.statusCode = code
+}
+
+// Write is just the collector for BatchResponse
+func (rec *BatchWriter) Write(b []byte) (int, error) {
+
+	var data interface{}
+	if err := json.Unmarshal(b, &data); err != nil {
+		return 0, err
+	}
+	if rec.statusCode != http.StatusOK {
+		rec.BResponse.ErrorCount++
+	}
+	requestID := ""
+	if header, ok := rec.headerMap["X-Request-Id"]; ok {
+		requestID = header[0]
+
+	}
+	bri := BatchResposeItem{Status: rec.statusCode, OrigResponse: data, Method: rec.method, URL: rec.url, RequestID: requestID}
+	rec.BResponse.Response = append(rec.BResponse.Response, bri)
+	rec.BResponse.EndedAt = time.Now()
+	return 0, nil
+}
+
+// Header returns header map
+func (rec *BatchWriter) Header() http.Header {
+	return rec.headerMap
+}
+
+// Request is the original request that is used for batching
+type Request struct {
```

Comment:
`Request` could be confused with the http package `Request`.
```suggestion
type BatchRequest struct {
```